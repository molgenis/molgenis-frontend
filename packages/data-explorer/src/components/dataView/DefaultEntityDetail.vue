<template>
  <div>
    <div class="d-flex m-2">
      <h2>{{ record.label || record.name }}</h2>
      <div v-if="isShop" class="ml-auto">
        <shopping-cart-button :id="record.id" />
      </div>
    </div>
    <ul class="list-group list-group-flush">
      <li v-for="(value, propertyName, index) in record" :key="index" class="list-group-item border-0">
        <h5>{{ propertyName }}</h5>
        <template v-if="!isObject(value)">
          <p>{{ value }}</p>
        </template>
        <template v-else-if="Array.isArray(value)">
          {{ value.map(v => v.label).join(', ') }}
        </template>
        <template v-else>
          <ul class="list-group list-group-flush">
            <li v-for="(refValue, refPropertyName, refIndex) in value" :key="refIndex" class="list-group-item border-0">
              <h5>{{ refPropertyName }}</h5>
              <p>{{ refValue }}</p>
            </li>
          </ul>
        </template>
      </li>
    </ul>
  </div>
</template>

<script>
import { mapState } from 'vuex'
import ShoppingCartButton from '../customViewComponents/ShoppingCartButton'

export default {
  name: 'DefaultEntityDetail',
  components: {
    ShoppingCartButton
  },
  props: {
    record: {
      type: Object,
      required: true
    },
    metaData: {
      type: Object,
      required: true
    }
  },
  computed: {
    ...mapState('explorer', {
      isShop: state => state.tableSettings.isShop
    }),
  },
  methods: {
    // doing this inline in the template breaks the vue code.
    isObject(value) {
      return typeof value === 'object'
    }
  }
}
</script>
