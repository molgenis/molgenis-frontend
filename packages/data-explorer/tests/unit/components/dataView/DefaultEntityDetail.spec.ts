import { createLocalVue, shallowMount } from '@vue/test-utils'
import DefaultEntityDetail from '@/components/dataView/DefaultEntityDetail.vue'
import Vuex from 'vuex'

const stubs = ['font-awesome-icon', 'router-link', 'b-tooltip']
const directives = { 'b-tooltip': () => {} }

const localVue = createLocalVue()
localVue.use(Vuex)

describe('DefaultEntityDetail.vue', () => {
  let wrapper

  const metaData = {
    label: 'my-entity-label',
    description: 'my-entity-desc',
    idAttribute: {
      name: 'id'
    }
  }

  const record = {
    id: 'abc-123',
    key1: 'val1',
    key2: ['val2', 'val3'],
    key3: {
      key4: 'val'
    }
  }
  let state: any

  beforeEach(async () => {
    state = {
      tableSettings: {
        isShop: false
      }
    }

    const explorer = { state, namespaced: true }
    const store = new Vuex.Store({ modules: { explorer } })

    wrapper = await shallowMount(DefaultEntityDetail, { stubs, directives, store, propsData: { metaData, record }, localVue })
  })

  it('exists', async () => {
    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can access the isShop property', () =>{
    expect(wrapper.vm.isShop).not.toBe(undefined)
  })
})
