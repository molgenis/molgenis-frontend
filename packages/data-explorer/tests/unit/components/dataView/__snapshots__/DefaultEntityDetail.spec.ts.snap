// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultEntityDetail.vue exists 1`] = `
<div>
  <div class="d-flex m-2">
    <h2></h2>
    <!---->
  </div>
  <ul class="list-group list-group-flush">
    <li class="list-group-item border-0">
      <h5>id</h5>
      <p>abc-123</p>
    </li>
    <li class="list-group-item border-0">
      <h5>key1</h5>
      <p>val1</p>
    </li>
    <li class="list-group-item border-0">
      <h5>key2</h5>
      ,
    </li>
    <li class="list-group-item border-0">
      <h5>key3</h5>
      <ul class="list-group list-group-flush">
        <li class="list-group-item border-0">
          <h5>key4</h5>
          <p>val</p>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
